crates/extrinsics/src/call.rs

```diff
@@ -16,278 +16,362 @@
-        match call_result.result {
-            Ok(_) => {
-                if !self.output_json {
-                    super::print_gas_required_success(call_result.gas_required);
-                }
-                // use user specified values where provided, otherwise use the estimates
-                let ref_time = self
-                    .gas_limit
-                    .unwrap_or_else(|| call_result.gas_required.ref_time());
-                let proof_size = self
-                    .proof_size
-                    .unwrap_or_else(|| call_result.gas_required.proof_size());
+    /// Estimates the gas required for a contract call without modifying the blockchain.
+    ///
+    /// This function provides a gas estimation for contract calls, considering the
+    /// user-specified values or using estimates based on a dry run. The estimated gas
+    /// weight is returned, or an error is reported if the estimation fails.
+    ///
+    /// Returns the estimated gas weight of type [`Weight`] for contract calls, or an
+    /// error.
+    pub async fn estimate_gas(&self) -> Result<Weight> {
+        match (self.gas_limit, self.proof_size) {
+            (Some(ref_time), Some(proof_size)) => {
                 Ok(Weight::from_parts(ref_time, proof_size))
             }
-            Err(ref err) => {
-                let object = ErrorVariant::from_dispatch_error(err, &client.metadata())?;
-                if self.output_json {
-                    Err(anyhow!("{}", serde_json::to_string_pretty(&object)?))
-                } else {
-                    name_value_println!("Result", object, MAX_KEY_COL_WIDTH);
-                    display_contract_exec_result::<_, MAX_KEY_COL_WIDTH>(&call_result)?;
-                    Err(anyhow!("Pre-submission dry-run failed. Use --skip-dry-run to skip this step."))
+            _ => {
+                let call_result = self.call_dry_run().await?;
+                match call_result.result {
+                    Ok(_) => {
+                        // use user specified values where provided, otherwise use the
+                        // estimates
+                        let ref_time = self
+                            .gas_limit
+                            .unwrap_or_else(|| call_result.gas_required.ref_time());
+                        let proof_size = self
+                            .proof_size
+                            .unwrap_or_else(|| call_result.gas_required.proof_size());
+                        Ok(Weight::from_parts(ref_time, proof_size))
+                    }
+                    Err(ref err) => {
+                        let object = ErrorVariant::from_dispatch_error(
+                            err,
+                            &self.client.metadata(),
+                        )?;
+                        Err(anyhow!("Pre-submission dry-run failed. Error: {}", object))
+                    }
                 }
             }
         }
     }
```

Comment:
This function is duplicate code in call and instantiate

Reply:
Well, the matching on the result could go into a helper just like that. Everything could go into a helper if `Exec` was a trait instead. But this might require a bit more careful design. Also it's not the end of the world for me . Let's not bike shed the PR forever. The PR is already a good move into the right direction, we can still improve from here.